### ISSUE TITLE
add support for apple-clang 12

### ISSUE THREAD

**Issue by AndreyMlashkin** [conan-io/conan#7635]
$ conan -v 
Conan version 1.28.1

Error log:

$ conan install zlib/1.2.11@ --build "*"
Auto detecting your dev setup to initialize the default profile (/Users/andreimalashkin/.conan/profiles/default)
Found apple-clang 12.0
Default settings
	os=Macos
	os_build=Macos
	arch=x86_64
	arch_build=x86_64
	compiler=apple-clang
	compiler.version=12.0
	compiler.libcxx=libc++
	build_type=Release
*** You can change them in /Users/andreimalashkin/.conan/profiles/default ***
*** Or override with -s compiler='other' -s ...s***

ERROR: Invalid setting '12.0' is not a valid 'settings.compiler.version' value.
Possible values are ['10.0', '11.0', '5.0', '5.1', '6.0', '6.1', '7.0', '7.3', '8.0', '8.1', '9.0', '9.1']
Read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-invalid-setting"

**Comment by memsharded** [comment]
The problem with AppleClang 12 is that it is not an official release yet. Please read comments in: https://github.com/conan-io/conan/pull/7601

Related: https://github.com/conan-io/conan/issues/7223

I would recommend to manually add 12.0 to your ``settings.yml``, and wait until the official release is done and then we could add it to our default settings. 

Thanks for reporting!

**Comment by AndreyMlashkin** [comment]
Do you know, if there is any way to install and use older apple-clang compiler?

**Comment by memsharded** [comment]
Hi @AndreyMlashkin 

Did you install the Beta 12 yourself? Did you have to remove the Stable XCode 11 before that? Maybe just uninstalling the beta is good enough. Some threads on the internet suggest just trashing the Xcode12 would be enough.

**Comment by AndreyMlashkin** [comment]
I've got it installed after I've done factory reset on my Mac

**Comment by memsharded** [comment]
So a Beta was automatically installed in your system after a factory reset???? This is quite unexpected from Apple (and to be honest, by other OSs).

Can you check if you have other XCode installed in parallel in your machine beside this beta?
I am going to ask for help to other OSX users.

**Comment by AndreyMlashkin** [comment]
I have only installed brew, python and conan and nothing else.
I don't have XCode installed on this machine.

**Comment by memsharded** [comment]
Ok, then it is quite unexpected that Conan found AppleClang12 there.

To develop C and C++, besides Conan you would need a few things:
- Compiler: You need to install XCode, that will contain the default compiler for OSX, which is AppleClang. Stable XCode 11 recommended.
- Most likely you need at least CMake. Many projects out there use CMake to build. It is possible to install and use CMake from Conan packages, but I would still recommend to start CMake at the system to start using Conan.

Then, remove the ``<userhome>/.conan`` folder, and start again. Check the output, it should detect AppleClang 11, and from there things should work.

Please try that and let me know how it goes.

**Comment by AndreyMlashkin** [comment]
I am still getting the same error after installing XCode.
Also I have changed the profile manually, but CMake can not detect other Clang versions:

Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=apple-clang
compiler.cppstd=14
compiler.libcxx=libc++
compiler.version=11.0
os=Macos
os_build=Macos
[options]
*:fPic=True
*:shared=False
[build_requires]
[env]


double-conversion/3.1.5: Copying sources to build folder
double-conversion/3.1.5: Building your package in /Users/andreimalashkin/.conan/data/double-conversion/3.1.5/_/_/build/e5cef9e9131bd7ddfe1fa83ed82c9d9e3e283fad
double-conversion/3.1.5: Generator cmake created conanbuildinfo.cmake
double-conversion/3.1.5: Calling build()
-- The C compiler identification is AppleClang 12.0.0.12000031
-- The CXX compiler identification is AppleClang 12.0.0.12000031
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan CMake Wrapper
-- Conan: called by CMake conan helper
-- Conan: called inside local cache
-- Conan: Adjusting output directories
-- Conan: Using cmake global configuration
-- Conan: Adjusting default RPATHs Conan policies
-- Conan: Adjusting language standard
-- Conan setting CPP STANDARD: 14 WITH EXTENSIONS OFF
-- Conan: Adjusting fPIC flag (ON)
CMake Error at conanbuildinfo.cmake:403 (message):
  Detected a mismatch for the compiler version between your conan profile
  settings and CMake:

  Compiler version specified in your conan profile: 11.0

  Compiler version detected in CMake: 12.0

  Please check your conan profile settings (conan profile show
  [default|your_profile_name])

  P.S.  You may set CONAN_DISABLE_CHECK_COMPILER CMake variable in order to
  disable this check.
Call Stack (most recent call first):
  conanbuildinfo.cmake:505 (conan_error_compiler_version)
  conanbuildinfo.cmake:588 (check_compiler_version)
  conanbuildinfo.cmake:142 (conan_check_compiler)
  CMakeLists.txt:6 (conan_basic_setup)


-- Configuring incomplete, errors occurred!
See also "/Users/andreimalashkin/.conan/data/double-conversion/3.1.5/_/_/build/e5cef9e9131bd7ddfe1fa83ed82c9d9e3e283fad/build_subfolder/CMakeFiles/CMakeOutput.log".
double-conversion/3.1.5: 
double-conversion/3.1.5: ERROR: Package 'e5cef9e9131bd7ddfe1fa83ed82c9d9e3e283fad' build failed
double-conversion/3.1.5: WARN: Build folder /Users/andreimalashkin/.conan/data/double-conversion/3.1.5/_/_/build/e5cef9e9131bd7ddfe1fa83ed82c9d9e3e283fad
ERROR: double-conversion/3.1.5: Error in build() method, line 43
	cmake = self._configure_cmake()
while calling '_configure_cmake', line 39
	cmake.configure(build_folder=self._build_subfolder)
	ConanException: Error 1 while executing cd '/Users/andreimalashkin/.conan/data/double-conversion/3.1.5/_/_/build/e5cef9e9131bd7ddfe1fa83ed82c9d9e3e283fad/build_subfolder' && cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE="Release" -DCMAKE_OSX_ARCHITECTURES="x86_64" -DCMAKE_OSX_SYSROOT="/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk" -DCONAN_CMAKE_CXX_STANDARD="14" -DCONAN_CMAKE_CXX_EXTENSIONS="OFF" -DCONAN_STD_CXX_FLAG="-std=c++14" -DCONAN_IN_LOCAL_CACHE="ON" -DCONAN_COMPILER="apple-clang" -DCONAN_COMPILER_VERSION="11.0" -DCONAN_CXX_FLAGS="-m64" -DCONAN_SHARED_LINKER_FLAGS="-m64" -DCONAN_C_FLAGS="-m64" -DCONAN_LIBCXX="libc++" -DBUILD_SHARED_LIBS="OFF" -DCMAKE_INSTALL_PREFIX="/Users/andreimalashkin/.conan/data/double-conversion/3.1.5/_/_/package/e5cef9e9131bd7ddfe1fa83ed82c9d9e3e283fad" -DCMAKE_INSTALL_BINDIR="bin" -DCMAKE_INSTALL_SBINDIR="bin" -DCMAKE_INSTALL_LIBEXECDIR="bin" -DCMAKE_INSTALL_LIBDIR="lib" -DCMAKE_INSTALL_INCLUDEDIR="include" -DCMAKE_INSTALL_OLDINCLUDEDIR="include" -DCMAKE_INSTALL_DATAROOTDIR="share" -DCONAN_CMAKE_POSITION_INDEPENDENT_CODE="ON" -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY="ON" -DCONAN_EXPORTED="1" -Wno-dev '/Users/andreimalashkin/.conan/data/double-conversion/3.1.5/_/_/build/e5cef9e9131bd7ddfe1fa83ed82c9d9e3e283fad'

**Comment by memsharded** [comment]
Maybe this could be an issue with CMake itself, if I understood correctly this: https://gitlab.kitware.com/cmake/cmake/-/issues/20893. 

Which version of CMake are you running?

**Comment by AndreyMlashkin** [comment]
$ cmake -version
cmake version 3.18.2

**Comment by memsharded** [comment]
Please @jgsogo @czoido @SSE4 could you help with this issue?

@AndreyMlashkin Please provide a bit more detail:

- It seems you are trying to install the https://conan.io/center/double-conversion/3.1.5, are you doing just ``conan install double-conversion/3.1.5@ --build``?
- Does it happen the same for other packages like zlib?


Thanks!

**Comment by SSE4** [comment]
the last error is:
```
Compiler version specified in your conan profile: 11.0

Compiler version detected in CMake: 12.0

```
I see your profile defines version 11.0, while the actual compiler version is 12.0. that's why CMake complains.

I think the steps should be:
1. add `12.0` version to the `settings.yml`, so it has lines:
```
    apple-clang: &apple_clang
        cppstd: [None, 98, 20, gnu98, 11, gnu11, 14, gnu14, 17, gnu17, 20, gnu20, 20]
        version: ["5.0", "5.1", "6.0", "6.1", "7.0", "7.3", "8.0", "8.1", "9.0", "9.1", "10.0", "11.0", "11.0.0"]

```

2. update conan profile to use `compiler.version=12.0`, so your profile would contain:
```
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=apple-clang
compiler.cppstd=14
compiler.libcxx=libc++
compiler.version=12.0
os=Macos
os_build=Macos
[options]
*:fPic=True
*:shared=False
[build_requires]
[env]


```

**Comment by AndreyMlashkin** [comment]
I have added compiler to settings.yml and now I can compile zlib and many other libs.

But it still fail with
conan install double-conversion/3.1.5@ --build missing

**Comment by jgsogo** [comment]
Hi, @AndreyMlashkin 

Please, share the error you get while building `double-conversion`, it is important to know if the error is originated by some Conan configuration or it is just because the library doesn't compile with that new apple compiler.

Thanks!

**Comment by memsharded** [comment]
Hi @SSE4 

It was not fully clear my request:

- I don't want to use AppleClang 12 beta, I would like to be using AppleClang 11, which is the latest stable one
- For some reason it seems AppleClang 12 beta is in a fresh system, even without installing XCode at all.
- The user didn't install AppleClang 12, and even after installing XCode and AppleClang 11, CMake refuses to detect it (it seems conan succeeded to detect it in the default profile)

It is very important that users by default will use the latest stable AppleClang11, and if this is not what is happening in OSX, we need to understand it and possibly change and add 12 to our default settings.yml.

**Comment by memsharded** [comment]
Quick question @AndreyMlashkin 

Which is your OS version? Are you in BigSur BETA?

**Comment by AndreyMlashkin** [comment]
Here is the MacOs version:
      System Version: macOS 10.15.6 (19G2021)
      Kernel Version: Darwin 19.6.0

**Comment by AndreyMlashkin** [comment]
I am able to compile the most of the recipes with apple-clang 12.
Please, add it to the default settings.yml

**Comment by jgsogo** [comment]
Hi, @AndreyMlashkin . XCode 12 is still a Beta release, the PR to add `apple-clang` 12 to the default `settings.yml` is already opened (https://github.com/conan-io/conan/pull/7601), but we will wait until it is a stable version. Binaries generated by the Beta might not be compatible with binaries generated by the stable version (if they change some default flag or implementation detail), that's the reason we prefer to wait until the stable version is released. 

Nevertheless, as you have already done, a Conan user can always override the `settings.yml` in their local installation, and Conan provides several ways to maintain a shared copy of that configuration in a team of developers.

Just a little bit of patience, Big Sur and XCode 12 will be ready before the end of the year 😀

**Comment by AndreyMlashkin** [comment]
Can I use CONAN_DISABLE_CHECK_COMPILER  to avoid this error? How can I activate it? 

I've tried to add this to my conan profile, but it didn't help

[env]
CONAN_DISABLE_CHECK_COMPILER=True

**Comment by memsharded** [comment]
Hi @AndreyMlashkin 

Nop, ``CONAN_DISABLE_CHECK_COMPILER`` is a cmake variable, not an enviornment variable. Define it in your ``CMakeLists.txt`` before calling ``conan_basic_setup()``.

**Comment by ssrobins** [comment]
Xcode 12 has been officially released :)

**Comment by memsharded** [comment]
Thanks for telling @ssrobins !

lets add it asap, at least in 1.30 and if there is another 1.29 patch release, we could include it.

**Comment by czoido** [comment]
Closed by https://github.com/conan-io/conan/pull/7722
